controllers/cache_controller.go

```diff
@@ -382,89 +388,140 @@ func (cl *CacheListener) CreateOrUpdate(data []byte) error {
-						return fmt.Errorf("unable to convert cache configuration from '%s' to '%s': %w", mime.ApplicationYaml, mediaType, err)
+				var template, templateName string
+				if cache.Spec.Template != "" {
+					cl.Log.Infof("Update Cache CR for '%s'\n%s", cache.Name, configYaml)
+					// Determinate the original user markup format and convert stream configuration to that format if required
+					mediaType := mime.GuessMarkup(cache.Spec.Template)
+					if mediaType == mime.ApplicationYaml {
+						template = configYaml
+					} else {
+						ispnClient, err := NewInfinispan(cl.Ctx, cl.Infinispan, cl.Kubernetes)
+						if err != nil {
+							return fmt.Errorf("unable to create Infinispan client: %w", err)
+						}
+						if template, err = ispnClient.Caches().ConvertConfiguration(configYaml, mime.ApplicationYaml, mediaType); err != nil {
+							return fmt.Errorf("unable to convert cache configuration from '%s' to '%s': %w", mime.ApplicationYaml, mediaType, err)
+						}
 					}
+				} else {
+					templateName = cache.Spec.TemplateName
 				}
-			} else {
-				templateName = cache.Spec.TemplateName
-			}
-			if cache.ObjectMeta.Annotations == nil {
-				cache.ObjectMeta.Annotations = make(map[string]string, 1)
-			}
-			controllerutil.AddFinalizer(cache, constants.InfinispanFinalizer)
-			cache.ObjectMeta.Annotations[constants.ListenerAnnotationGeneration] = strconv.FormatInt(cache.GetGeneration()+1, 10)
-			cache.Spec = v2alpha1.CacheSpec{
-				Name:         cacheName,
-				ClusterName:  cl.Infinispan.Name,
-				Template:     template,
-				TemplateName: templateName,
+
+				if cache.ObjectMeta.Annotations == nil {
+					cache.ObjectMeta.Annotations = make(map[string]string, 1)
+				}
+				controllerutil.AddFinalizer(cache, constants.InfinispanFinalizer)
+				cache.ObjectMeta.Annotations[constants.ListenerAnnotationGeneration] = strconv.FormatInt(cache.GetGeneration()+1, 10)
+				cache.Spec = v2alpha1.CacheSpec{
+					Name:         cacheName,
+					ClusterName:  cl.Infinispan.Name,
+					Template:     template,
+					TemplateName: templateName,
+				}
+				return nil
+			})
+			if err == nil {
+				break
 			}
-			return nil
-		})
 
-		if err == nil {
-			break
+			if !errors.IsConflict(err) {
+				return fmt.Errorf("unable to Update Cache CR '%s': %w", cache.Name, err)
+			}
+			cl.Log.Errorf("Conflict encountered on Cache CR '%s' update. Retry %d..%d", cache.Name, i, maxRetries)
 		}
-
-		if !errors.IsConflict(err) {
-			return fmt.Errorf("unable to CreateOrUpdate Cache CR: %w", err)
+		if err != nil {
+			return fmt.Errorf("unable to Update Cache CR %s after %d attempts", cache.Name, maxRetries)
 		}
-		cl.Log.Errorf("Conflict encountered on Cache update. Retry %d..%d", i, maxRetries)
 	}
+	return nil
+}
 
-	if err != nil {
-		return fmt.Errorf("unable to CreateOrPatch Cache CR %s after %d attempts", cacheCrName, maxRetries)
+func (cl *CacheListener) findExistingCacheCR(cacheName, clusterName string) (*v2alpha1.Cache, error) {
```

Comment:
I did consider that, however there are a few issues:

1. We need to apply the labels to user created Cache CRs. Easily solvable
2. How to handle label length restrictions. Max label length is 63 chars and we can't have special characters

2 is tricky as we don't have this length limit on the cache name and cluster name elsewhere, so it could create an inconsistent experience if the length restrictions only apply to specific cases.

Reply:
Hm... Okay, with those restrictions it doesn't make that much sense. I was thinking there are no restrictions on labels. We can eventually improve on this later on

Reply:
We could have a hybrid approach where we set `org.infinispan/clusterName` and then just cycle through caches belonging to that cluster.

Reply:
We could also hash a cache name for the label, obviously it could potentially result in several results that we'd have to cycle through. Not sure if it's worth the effort.

Reply:
I like that idea :+1: 

As we require this downstream sooner rather than later, my preference would be to merge this and then we can improve the performance via such means in a follow up PR.